database/pebble/iterator.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package pebble
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/cockroachdb/pebble"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/ava-labs/avalanchego/database"
+)
+
+var (
+	_ database.Iterator = (*iter)(nil)
+
+	errCouldntGetValue = errors.New("couldnt get iterator value")
+	errNoPointKey      = errors.New("iterator has no point key")
+)
+
+type iter struct {
+	// [lock] ensures that only one goroutine can access [iter] at a time.
+	// Note that [Database.Close] calls [iter.Release] so we need [lock] to ensure
+	// that the user and [Database.Close] don't execute [iter.Release] concurrently.
+	// Invariant: [Database.lock] is never grabbed while holding [lock].
+	lock sync.Mutex
+
+	db   *Database
+	iter *pebble.Iterator
+
+	initialized bool
+	closed      bool
+	err         error
+
+	hasNext bool
+	nextKey []byte
+	nextVal []byte
+}
+
+// Must not be called with [db.lock] held.
+func (it *iter) Next() bool {
+	it.db.lock.RLock()
+	dbClosed := it.db.closed
+	it.db.lock.RUnlock()
+
+	it.lock.Lock()
+	defer it.lock.Unlock()
+
+	switch {
+	case it.closed || dbClosed:
```

Comment:
`dbClosed` is not needed here - because `dbClosed` implies `it.closed`. This is guaranteed because `db.Close` grabs all iterator locks before closing the underlying database and releases them.
```suggestion
	it.lock.Lock()
	defer it.lock.Unlock()

	switch {
	case it.closed:
```